lib/controller/controller.go

```diff
@@ -129,67 +97,131 @@ type ProvisionController struct {
 
-	// Interval between retries when we create a PV object for a provisioned volume.
-	createProvisionedPVInterval time.Duration
-
-	// Identity of this controller, generated at creation time and not persisted
-	// across restarts. Useful only for debugging, for seeing the source of
-	// events. controller.provisioner may have its own, different notion of
-	// identity which may/may not persist across restarts
-	identity types.UID
+	failedProvisionThreshold, failedDeleteThreshold int
+	// Map of failed claims to provisions/volumes to deletes
+	failedProvisionStats, failedDeleteStats           map[types.UID]int
+	failedProvisionStatsMutex, failedDeleteStatsMutex *sync.Mutex
 
 	// Parameters of leaderelection.LeaderElectionConfig. Leader election is for
 	// when multiple controllers are running: they race to lock (lead) every PVC
 	// so that only one calls Provision for it (saving API calls, CPU cycles...)
-	// Descriptions of each can be found in the leaderelection package and
-	// reproduced above (DefaultLeaseDuration...)
 	leaseDuration, renewDeadline, retryPeriod, termLimit time.Duration
-
 	// Map of claim UID to LeaderElector: for checking if this controller
 	// is the leader of a given claim
-	leaderElectors map[types.UID]*leaderelection.LeaderElector
-	mapMutex       *sync.Mutex
+	leaderElectors      map[types.UID]*leaderelection.LeaderElector
+	leaderElectorsMutex *sync.Mutex
+}
+
+const (
+	defaultResyncPeriod                  = 15 * time.Second
+	defaultExponentialBackOffOnError     = true
+	defaultCreateProvisionedPVRetryCount = 5
+	defaultCreateProvisionedPVInterval   = 10 * time.Second
+	defaultFailedProvisionThreshold      = 5
+	defaultFailedDeleteThreshold         = 5
+	defaultLeaseDuration                 = 15 * time.Second
+	defaultRenewDeadline                 = 10 * time.Second
+	defaultRetryPeriod                   = 2 * time.Second
+	defaultTermLimit                     = 30 * time.Second
+)
 
-	// Threshold for max number of retries on failure of provisioner
-	failedRetryThreshold int
+// ResyncPeriod is how often the controller relists PVCs, PVs, & storage
+// classes. OnUpdate will be called even if nothing has changed, meaning failed
+// operations may be retried on a PVC/PV every resyncPeriod regardless of
+// whether it changed. Defaults to 15 seconds.
+func ResyncPeriod(resyncPeriod time.Duration) func(*ProvisionController) error {
+	return func(c *ProvisionController) error { c.resyncPeriod = resyncPeriod; return nil }
+}
 
-	// Map of failed claims
-	failedClaimsStats      map[types.UID]int
-	failedClaimsStatsMutex *sync.Mutex
+// ExponentialBackOffOnError determines whether to exponentially back off from
+// failures of Provision and Delete. Defaults to true.
+func ExponentialBackOffOnError(exponentialBackOffOnError bool) func(*ProvisionController) error {
+	return func(c *ProvisionController) error {
+		c.runningOperations = goroutinemap.NewGoRoutineMap(exponentialBackOffOnError)
```

Comment:
It's so verbose I am considering not throwing errors if called in controller runtime at all. It should be understood that these are functional options and not something to call during runtime IMO, I will think baout it some more.